### ISSUE TITLE
Mixtral past_key_values and output_router_logits incompatible

### ISSUE THREAD

**Issue by sorgfresser** [huggingface/transformers#30731]
### System Info

transformers==4.40.2
Python 3.11.8

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers import MixtralConfig, MixtralForCausalLM, AutoTokenizer
import torch
# Initializing a smaller version of Mixtral for faster execution
configuration = MixtralConfig(
    hidden_size=256,
    intermediate_size=896,
    num_hidden_layers=8,
    num_attention_heads=8,
    num_key_value_heads=8,
    num_local_experts=4,
    num_experts_per_tok=1,
)

model = MixtralForCausalLM(configuration)
tokenizer = AutoTokenizer.from_pretrained("mistralai/Mixtral-8x7B-v0.1")
prompt = "This is a test"
tokenized = tokenizer(prompt, return_tensors="pt")
output = model(**tokenized, output_router_logits=True)
key_values = output.past_key_values
logits = output.logits
next_token_logits = logits[..., -1, :]
# Softmax
softmaxed = torch.nn.functional.softmax(next_token_logits, dim=-1)
# Sample
sampled = torch.multinomial(softmaxed.squeeze(), num_samples=1)
ids = sampled.item()

attention_mask = torch.cat([tokenized["attention_mask"], torch.tensor([[1]])], dim=-1)
next_output = model(
    torch.tensor([[ids]]),
    attention_mask=attention_mask,
    past_key_values=key_values,
    output_router_logits=True
)
```

### Expected behavior

It seems that this is the same underlying issue as in #29087 - I would expect `past_key_values` to work with `output_router_logits`. 
So what happens?
1. Without past key values (and with multiple input ids) the `all_router_logits` has the proper sequence length, thus in `load_balancing_loss_func` this `num_hidden_layers = concatenated_gate_logits.shape[0] // (batch_size * sequence_length)` correctly evaluates the number of hidden layers.
2. If past key values are used, `all_router_logits` has a sequence length of 1, but since the attention mask is still the whole sequence (from which the `sequence_length` is inferred) the hidden layers evaluate to a small value or 0, leading to the same error as in #29087

Instead, I would like the `load_balancing_loss_func` to be able to deal with a case where the `gate_logits` passed are of shape `[batch_size X 1, num_experts]` instead of `[batch_size X sequence_length, num_experts]`.

**Comment by cw235** [comment]
It looks like you are encountering an issue related to using `past_key_values` with `output_router_logits` in your customized implementation with the Mixtral model from Hugging Face's Transformers library.

In the provided code snippet, you are using `past_key_values` along with `output_router_logits` to generate next token predictions. However, you have observed that when `past_key_values` are used, the `all_router_logits` has a sequence length of 1, which affects the calculation of the number of hidden layers in your custom `load_balancing_loss_func`.

The issue you are facing might be related to how the `past_key_values` are handled and how they interact with the `output_router_logits`. `past_key_values` are used to speed up generation by reusing the intermediate calculations from the previous time steps, but there may be a mismatch in how the sequences are processed when using `past_key_values`.

One possible suggestion to address this issue is to ensure that the sequence length is correctly maintained and propagated when using `past_key_values`. You may need to adjust the handling of `past_key_values` and the attention mask to correctly account for the sequence length when calculating the number of hidden layers.

Additionally, you may consider checking the documentation for the Mixtral model or reaching out to the Hugging Face community or the Transformers library maintainers for further guidance and support in addressing this specific issue with `past_key_values` and `output_router_logits`.

If you require further assistance or clarification on any specific aspect of your implementation, feel free to provide more details, and I'll do my best to help!

**Comment by ArthurZucker** [comment]
Hey! The generate function is not supposed to work for training. That is why we don't test past key values and output router logits. Though it's actually not that incompatible (you could want to look at the distribution of the router logits during generation). 
Do you want to open a PR for a fix?

**Comment by csking101** [comment]
Hi, could I take this up? @ArthurZucker

**Comment by ArthurZucker** [comment]
Sur feel free to open a PR!

**Comment by csking101** [comment]
I am having trouble figuring out how to go about the `[batch_size X 1, num_experts]` case. I am thinking of creating a condition for this in the `load_balancing_loss_func`, such as, `is_single_token = gate_logits[0].shape[0] == 1`. After that, I will modify the code in the following way, so that the shapes during the multiplication are valid:

```
        if not is_single_token:
            expert_attention_mask = (
                attention_mask[None, :, :, None, None]
                .expand((num_hidden_layers, batch_size, sequence_length, top_k, num_experts))
                .reshape(-1, top_k, num_experts)
                .to(compute_device)
            )
        else:
            expert_attention_mask = (
                attention_mask[None, :, -1:, None, None]
                .expand((len(gate_logits), 1,1, top_k, num_experts))
                .reshape(-1, top_k, num_experts)
                .to(compute_device)
            )
```
As you can see here, I think one approach is to retain the attention score of that particular token itself (the one that we are adding), and get rid of the other scores in the sequence. This is because, if we do retain the other ones, then `expert_attention_mask`'s shape will not match that of  `expert_mask`. As a consequence, for this case, we will not need to calculate the `num_hidden_layers` (as it is a tiny value as mentioned by @sorgfresser) , because we will just get the batch size directly from `gate_logits`.  Similarly, we will need a modification for `router_per_expert_attention_mask`:
```
if not is_single_token:
            router_per_expert_attention_mask = (
                attention_mask[None, :, :, None]
                .expand((num_hidden_layers, batch_size, sequence_length, num_experts))
                .reshape(-1, num_experts)
                .to(compute_device)
            )
        else:
            router_per_expert_attention_mask = (
                attention_mask[None, :, -1:, None]
                .expand((len(gate_logits), 1, 1, num_experts))
                .reshape(-1, num_experts)
                .to(compute_device)
            )
```

This approach is what I have come up with, I hope you all can verify the correctness of the same, and point out if I have made any mistakes. The PR is [here](https://github.com/huggingface/transformers/pull/34707).

**Comment by tkj666** [comment]
I think a better way to handle this is to get `num_hidden_layers` from `len(gate_logits)` and `batch_size` from `attention_mask.shape[0]`, then calculate actual `sequence_length` to slice `attention_mask`

The PR is #40908
